opendevin/server/session/agent.py

```diff
@@ -0,0 +1,104 @@
+from typing import Optional
+
+from agenthub.codeact_agent.codeact_agent import CodeActAgent
+from opendevin.controller import AgentController
+from opendevin.controller.agent import Agent
+from opendevin.controller.state.state import State
+from opendevin.core.config import config
+from opendevin.core.logger import opendevin_logger as logger
+from opendevin.core.schema import ConfigType
+from opendevin.events.stream import EventStream
+from opendevin.llm.llm import LLM
+from opendevin.runtime import DockerSSHBox
+from opendevin.runtime.e2b.runtime import E2BRuntime
+from opendevin.runtime.runtime import Runtime
+from opendevin.runtime.server.runtime import ServerRuntime
+
+
+class AgentSession:
+    """Represents a session with an agent.
+
+    Attributes:
+        controller: The AgentController instance for controlling the agent.
+    """
+
+    sid: str
+    event_stream: EventStream
+    controller: Optional[AgentController] = None
+    runtime: Optional[Runtime] = None
+
+    def __init__(self, sid):
+        """Initializes a new instance of the Session class."""
+        self.sid = sid
+        self.event_stream = EventStream(sid)
+
+    async def start(self, start_event: dict):
+        """Starts the agent session.
+
+        Args:
+            start_event: The start event data (optional).
+        """
+        await self._create_runtime()
+        await self._create_controller(start_event)
+
+    async def _create_runtime(self):
```

Comment:
Do we need to check if self.runtime is not None, and close it if already exists. Just like what we do in `_create_controller`

Reply:
Modified this so that we throw errors if someone tries to reuse the same session w/ a new runtime or controller